libpimeval/src/pimPerfEnergyModels.cpp

```diff
@@ -125,107 +205,260 @@ pimParamsPerf::getPerfEnergyBitSerial(PimDeviceEnum deviceType, PimCmdEnum cmdTy
+      msRuntime = m_tW + m_tL * maxBytesPerRegion; // for one pass
+      msRuntime *= numPass;
+      mjEnergy = (m_eAP + (m_tL * maxBytesPerRegion)) * numCore * numPass;
       mjEnergy += m_pBChip * m_numChipsPerRank * numRanks * msRuntime;
       break;
     }
     default:
-      std::printf("PIM-Warning: Perf energy model not available for PIM command %s\n", pimCmd::getName(cmdType, "").c_str());
+      assert(0);
+  }
+
+  return pimNS::perfEnergy(msRuntime, mjEnergy);
+}
+
+//! @brief  Perf energy model of bit-serial PIM for rotate
+pimNS::perfEnergy
+pimPerfEnergyBitSerial::getPerfEnergyForRotate(PimCmdEnum cmdType, const pimObjInfo& obj) const
+{
+  PimDeviceEnum simTarget = pimSim::get()->getSimTarget();
+  double msRuntime = 0.0;
+  double mjEnergy = 0.0;
+  unsigned numPass = obj.getMaxNumRegionsPerCore();
+  unsigned bitsPerElement = obj.getBitsPerElement();
+  unsigned numRegions = obj.getRegions().size();
+  // boundary handling
+  pimNS::perfEnergy perfEnergyBT = getPerfEnergyForBytesTransfer(cmdType, numRegions * bitsPerElement / 8);
+
+  switch (simTarget) {
+    case PIM_DEVICE_BITSIMD_V:
+    case PIM_DEVICE_BITSIMD_V_AP:
+      // rotate within subarray:
+      // For every bit: Read row to SA; move SA to R1; Shift R1; Move R1 to SA; Write SA to row
+      msRuntime = (m_tR + 3 * m_tL + m_tW) * bitsPerElement; // for one pass
+      msRuntime *= numPass;
+      mjEnergy = (m_eAP + 3 * m_eL) * bitsPerElement * numPass; // for one pass
+      msRuntime += 2 * perfEnergyBT.m_msRuntime;
+      mjEnergy += 2 * perfEnergyBT.m_mjEnergy;
       break;
-    }
-    break;
+    case PIM_DEVICE_SIMDRAM:
+      // todo
+      break;
+    case PIM_DEVICE_BITSIMD_H:
+      // rotate within subarray:
+      // For every bit: Read row to SA; move SA to R1; Shift R1 by N steps; Move R1 to SA; Write SA to row
+      // TODO: separate bank level and GDL
+      // TODO: energy unimplemented
+      msRuntime = (m_tR + (bitsPerElement + 2) * m_tL + m_tW); // for one pass
+      msRuntime *= numPass;
+      mjEnergy = (m_eAP + (bitsPerElement + 2) * m_eL) * numPass;
+      msRuntime += 2 * perfEnergyBT.m_msRuntime;
+      mjEnergy += 2 * perfEnergyBT.m_mjEnergy;
+      break;
+    default:
+      assert(0);
   }
-  case PIM_DEVICE_BANK_LEVEL:
+
+  return pimNS::perfEnergy(msRuntime, mjEnergy);
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+// MODEL: Fulcrum
+///////////////////////////////////////////////////////////////////////////////
+
+//! @brief  Perf energy model of Fulcrum for func1
+pimNS::perfEnergy
+pimPerfEnergyFulcrum::getPerfEnergyForFunc1(PimCmdEnum cmdType, const pimObjInfo& obj) const
+{
+  double msRuntime = 0.0;
+  double mjEnergy = 0.0;
+  unsigned numPass = obj.getMaxNumRegionsPerCore();
+  unsigned bitsPerElement = obj.getBitsPerElement();
+  unsigned numCores = obj.getNumCoresUsed();
+  unsigned numRanks = pimSim::get()->getNumRanks();
```

Comment:
These are going to be duplicated. I wonder if we should add these in the constructor to avoid duplication?

Reply:
Yes. A params class is needed since pimDevice is not fully constructed during perf energy model ctor.